# Ago-Seq

Exploring short DNA fragmants co-purified with Ago proteins from *E. coli* cells.

This repository contains a set of bash and python scripts which were used for Ago-Seq data analysis and visualization. 


######################

## Bed_to_wig_converter.py

Script takes BED files with coverage depth and converts them to WIG format.

**Requirements:** Python 2 or 3

**Input:** BED files

**Output:** WIG files

######################

## Calculate_FPKM.py

Script takes WIG files with coverage depth and a dictionary with total number of sequenced reads per fule. 
Usinf these data, it converts coverage depth tracks to RPKM tracks.

**Requirements:** Python 3

**Input:** WIG files, number of reads per track

**Output:** WIG files with RPKM values

######################

## Average_wig_files.py

Takes a set of WIG files (organized as a dictionary) and computes by-position average WIG. Returns correlation matrix 
as a heatmap of pair-wize correlations between genomic tracks.

**Requirements:** Python 2 or 3

**Input:** WIG files to be averaged

**Output:** Averaged WIG file, correlation matrix heatmap (PNG)

######################

## FE_over_US_GB_DS_binning_and_statistics.py

Takes wig tracks of different continuous genomic features (Ago-Seq data for F and R strands, etc.). Computes signal over TUs upstream (US),
downstream (DS) and over TUs bodies (GB). Script is dedicated for noisy data. 
To handle the data it performes data binning for smoothing. Also it keeps signal data for all TUs to contstruct confidential interval.
To plot different features together (e.g. ChIP-Seq data from different expriments or conditions) it normalizes the data.

**Requirements:** Python 3

**Input:** Files with signal data (WIG), genome annotation (GFF or BroadPeak), regions to be omitted (BroadPeak), size of a bin

**Output:** WIG files with metagene signal over all TUs, TAB files with average signal for each of TUs, 
metagen plot of average signal over all TUs, bar plot of average signal in US, DS regions and in TUs bodies

######################

## FE_over_intergenic_regions.py

Takes wig tracks of different genome features (Ago-Seq data for F and R strands, etc.). Computes signal in intergenic regions (IGRs, provides as BroadPeak or GFF file), in respect to 
orientation of adjacent genes: ++, +-, -+ or --. 

**Requirements:** Python 3

**Input:** Files with signal data (WIG), IGRs annotation (GFF or BroadPeak), regions to be omitted (e.g. deletions, BroadPeak)

**Output:** WIG files with metagene signal over IGRs classified by adjacent genes orientation, TAB files with average signal for the each of IGRs, 
metaplot of average signal over IGR groups, histogram of the signal over IGRs

######################

## Plot_signal_over_intergenic_regions.py

Takes signal over intergenic regions in WIG format generated by FE_over_intergenic_regions.py and plots metagene signal over IGRs.

**Requirements:** Python 3

**Input:** WIG files with metagene signal over IGR sets

**Output:** Plots with metagene signal

######################

## GCSs_as_anchors.py

Takes wig tracks of different genome features (Ago-Seq data for F and R strands, etc.). Computes metasignal relative to some symmetrical anchors (e.g., GCSs).

**Requirements:** Python 3

**Input:** Files with signal data (WIG), anchors positions (.txt), regions to be omitted (e.g. deletions, BroadPeak)

**Output:** WIG files with metasignal over anchors, TAB files with average signal for the each of anchors, 
metaplot of average signal over anchors

######################

## Chi_as_anchors.py

Takes wig tracks of different genome features (Ago-Seq data for F and R strands, etc.).
Computes metasignal relative to asymmetrical anchors (e.g., Chi-sites).

**Requirements:** Python 3

**Input:** Files with signal data (WIG), motif sequence, regions to be omitted (e.g. deletions, BroadPeak)

**Output:** WIG files with metasignal over anchors, TAB files with average signal for the each of anchors, 
TAB files with a full (not averaged) signal for the each of anchors, metaplot of average signal over anchors

######################

## Plot_signal_over_symmetric_anchors.py

Takes signal over symmetric anchors in WIG format generated by GCSs_as_anchors.py and plots metagene signal over the anchor groups.

**Requirements:** Python 3

**Input:** WIG files with metagene signal over anchors sets

**Output:** Metaplots

######################

## Plot_signal_over_asymmetric_anchors.py

Takes signal over asymmetric anchors in WIG format generated by Chi_as_anchors.py and plots metagene signal over the anchor groups.

**Requirements:** Python 3

**Input:** WIG files with metagene signal over anchors sets

**Output:** Metaplots

######################

## Barplot_signal_over_asymmetric_anchors.py

Takes signal over asymmetric anchors in generated by Chi_as_anchors.py and statistically compares
signal positioned relative to the anchors.

**Requirements:** Python 3

**Input:** Files with metagene signal over anchors sets

**Output:** Barplots, t-test results

######################

## Chi_as_anchors_GCSs_downstream_vs_others.py

Takes wig tracks of different genome features (Ago-Seq data for F and R strands, etc.).
Classifies asymmetrical anchors 1 (e.g., Chi-sites) by their association with symmetrical anchors 2 (e.g., GCSs).
Groups anchors 1: anchors 1 adjacent to the upstream regions of anchors 2 and other anchors 1. 
Computes metasignal relative to groups of anchors 1 and statistically compares
signal positioned relative to the anchors.

**Requirements:** Python 3

**Input:** Files with signal data (WIG), motif sequence, regions to be omitted (e.g. deletions, BroadPeak), anchor 1 sequence, anchor 2 coordinates (bed-like)

**Output:** WIG files with metasignal over anchors, TAB files with average signal for the each of anchors, 
TAB files with a full (not averaged) signal for the each of anchors, metaplot of average signal over anchors,
Barplots, t-test results

######################

## Transfer_TUs_MG1655_to_BL21.py

Takes coordinates of transcription units in the E. coli MG1655 genome and translates them to the desired reference genome (e.g., NC_012971.2) using blastn.

**Requirements:** Python 3, blast+

**Input:** Reference genome 1 (FASTA), Reference genome 2 (FASTA), TUs coordinates (txt), regions to be omitted (e.g. deletions, BroadPeak)

**Output:** BroadPeak and bed files with translated intervals

######################

## Growth_curve_plot.py

Takes culture growth curve data in a form of xlsx table containing series of time-points and corresponding optical density of a culture. 
Plots mean and confidential intervals if several replicates were provided.

**Requirements:** Python 3

**Input:** Excell table with time-points and optical density measurments

**Output:** Growth curve plot
